python/cudf/cudf/tests/series/test_constructors.py

```diff
@@ -640,6 +643,195 @@ def test_series_arrow_decimal_types_roundtrip(pa_type):
         assert_eq(pdf, gdf)
 
 
+@pytest.mark.parametrize("module", ["cupy", "numba"])
+def test_cuda_array_interface_interop_in(
+    numeric_and_temporal_types_as_str, module
+):
+    if module == "cupy":
+        module_constructor = cp.array
+        if numeric_and_temporal_types_as_str.startswith(
+            "datetime"
+        ) or numeric_and_temporal_types_as_str.startswith("timedelta"):
+            pytest.skip(
+                f"cupy doesn't support {numeric_and_temporal_types_as_str}"
+            )
+    elif module == "numba":
+        module_constructor = numba.cuda.to_device
+
+    np_data = np.arange(10).astype(numeric_and_temporal_types_as_str)
+    module_data = module_constructor(np_data)
+
+    pd_data = pd.Series(np_data)
+    # Test using a specific function for __cuda_array_interface__ here
+    cudf_data = cudf.Series(module_data)
+
+    assert_eq(pd_data, cudf_data)
+
+    gdf = cudf.DataFrame()
+    gdf["test"] = module_data
+    pd_data.name = "test"
+    assert_eq(pd_data, gdf["test"])
+
+
+@pytest.mark.parametrize("module", ["cupy", "numba"])
+def test_cuda_array_interface_interop_out(
+    numeric_and_temporal_types_as_str, module
+):
+    if module == "cupy":
+        module_constructor = cp.asarray
+
+        def to_host_function(x):
+            return cp.asnumpy(x)
+    elif module == "numba":
+        module_constructor = numba.cuda.as_cuda_array
+
+        def to_host_function(x):
+            return x.copy_to_host()
+
+    np_data = np.arange(10).astype(numeric_and_temporal_types_as_str)
+    cudf_data = cudf.Series(np_data)
+    assert isinstance(cudf_data.__cuda_array_interface__, dict)
+
+    module_data = module_constructor(cudf_data)
+    got = to_host_function(module_data)
+
+    expect = np_data
+
+    assert_eq(expect, got)
+
+
+def test_cuda_array_interface_interop_out_masked(
+    numeric_and_temporal_types_as_str,
+):
+    np_data = np.arange(10).astype("float64")
+    np_data[[0, 2, 4, 6, 8]] = np.nan
+
+    cudf_data = cudf.Series(np_data).astype(numeric_and_temporal_types_as_str)
+    cai = cudf_data.__cuda_array_interface__
+    assert isinstance(cai, dict)
+    assert "mask" in cai
```

Comment:
This test originally tried having cupy/numba accept a `__cuda_array_interface__` with a mask, which they still both don't support. I'm open to better ideas/assertions to make about this mask without using another library that would raise an error.